In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#create path
path = r'C:\Users\jacob\Documents\Data Analytics\Data Immersion\Achievement 6\Energy_GDP_Project'

In [3]:
#import renewables data
renew = pd.read_csv(os.path.join(path, '02 data', 'original data', 'energy_from_renewables.csv'))

In [4]:
renew.head()

country  year       combined key  %_energy_renewables
0  Afghanistan  1990  Afghanistan, 1990                  NaN
1  Afghanistan  1991  Afghanistan, 1991                  NaN
2  Afghanistan  1992  Afghanistan, 1992                  NaN
3  Afghanistan  1993  Afghanistan, 1993                  NaN
4  Afghanistan  1994  Afghanistan, 1994                  NaN

In [5]:
renew.shape

(6916, 4)

In [6]:
renew.describe()

year  %_energy_renewables
count  6916.000000          4761.000000
mean   2002.500000             2.480903
std       7.500542             5.409203
min    1990.000000             0.000000
25%    1996.000000             0.000000
50%    2002.500000             0.411142
75%    2009.000000             2.080273
max    2015.000000            65.443742

In [7]:
#check for missing values
renew.isnull().sum()

country                   0
year                      0
combined key              0
%_energy_renewables    2155
dtype: int64

In [8]:
#check for duplicated values
renew.duplicated().value_counts()

False    6916
Name: count, dtype: int64

In [9]:
#group data by country and then look for missing values
renew_grouped = renew.groupby('country')['%_energy_renewables'].value_counts(dropna = False)

In [10]:
renew_grouped.head(150)

country                      %_energy_renewables
Afghanistan                  NaN                    26
Africa Eastern and Southern  2.818460                1
                             1.995052                1
                             0.468126                1
                             0.285685                1
                                                    ..
Australia                    0.546101                1
                             0.586094                1
                             0.586857                1
                             0.712996                1
Austria                      2.279749                1
Name: count, Length: 150, dtype: int64

There are no duplicates within this data set. However, there are 2155, missing vlaues. We scanning the data most of the values are from countries where no data has been collected. there are a handful of ocuntreis that are missing a few values. to combat this we will create a new columnm that will created a weighted average that will fill in missing values for countries that we do have data for. Then we will impute 0 for all of the other missing values. Then we will remove the original energy from renewables column and change the weighted avergage column into the energy from renewables column

In [11]:
#create a weighted average column to fill in missing data values
renew ['weighted_avg'] = renew.groupby('country')['%_energy_renewables'].transform(lambda x : x.fillna(x.mean()))

In [12]:
renew.head(150)

country  year          combined key  %_energy_renewables  \
0       Afghanistan  1990     Afghanistan, 1990                  NaN   
1       Afghanistan  1991     Afghanistan, 1991                  NaN   
2       Afghanistan  1992     Afghanistan, 1992                  NaN   
3       Afghanistan  1993     Afghanistan, 1993                  NaN   
4       Afghanistan  1994     Afghanistan, 1994                  NaN   
..              ...   ...                   ...                  ...   
145  American Samoa  2005  American Samoa, 2005                  NaN   
146  American Samoa  2006  American Samoa, 2006                  NaN   
147  American Samoa  2007  American Samoa, 2007                  NaN   
148  American Samoa  2008  American Samoa, 2008                  NaN   
149  American Samoa  2009  American Samoa, 2009                  NaN   

     weighted_avg  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
..            ...  
145           NaN  
146           NaN  
147           NaN  
148           NaN  
149           NaN  

[150 rows x 5 columns]

In [13]:
#check missing values
renew.isnull().sum()

country                   0
year                      0
combined key              0
%_energy_renewables    2155
weighted_avg           2080
dtype: int64

In [14]:
#fill remainder missing values with zero
renew['weighted_avg'].fillna(0, inplace = True)

In [15]:
renew.head(105)

country  year       combined key  %_energy_renewables  weighted_avg
0    Afghanistan  1990  Afghanistan, 1990                  NaN           0.0
1    Afghanistan  1991  Afghanistan, 1991                  NaN           0.0
2    Afghanistan  1992  Afghanistan, 1992                  NaN           0.0
3    Afghanistan  1993  Afghanistan, 1993                  NaN           0.0
4    Afghanistan  1994  Afghanistan, 1994                  NaN           0.0
..           ...   ...                ...                  ...           ...
100      Albania  2012      Albania, 2012                  0.0           0.0
101      Albania  2013      Albania, 2013                  0.0           0.0
102      Albania  2014      Albania, 2014                  0.0           0.0
103      Albania  2015      Albania, 2015                  0.0           0.0
104      Algeria  1990      Algeria, 1990                  0.0           0.0

[105 rows x 5 columns]

In [16]:
#check for missing values in weighted average column
renew.isnull().sum()

country                   0
year                      0
combined key              0
%_energy_renewables    2155
weighted_avg              0
dtype: int64

In [17]:
#drop energy_renewables column
renew.drop(columns = '%_energy_renewables', inplace = True)

In [18]:
renew.head()

country  year       combined key  weighted_avg
0  Afghanistan  1990  Afghanistan, 1990           0.0
1  Afghanistan  1991  Afghanistan, 1991           0.0
2  Afghanistan  1992  Afghanistan, 1992           0.0
3  Afghanistan  1993  Afghanistan, 1993           0.0
4  Afghanistan  1994  Afghanistan, 1994           0.0

In [19]:
#rename weighted average column to be new energy from renewablaes column
renew.rename(columns = {'weighted_avg' : '%_energy_renewables'}, inplace = True)

In [20]:
renew.head()

country  year       combined key  %_energy_renewables
0  Afghanistan  1990  Afghanistan, 1990                  0.0
1  Afghanistan  1991  Afghanistan, 1991                  0.0
2  Afghanistan  1992  Afghanistan, 1992                  0.0
3  Afghanistan  1993  Afghanistan, 1993                  0.0
4  Afghanistan  1994  Afghanistan, 1994                  0.0

In [21]:
#check for missing values
renew.isnull().sum()

country                0
year                   0
combined key           0
%_energy_renewables    0
dtype: int64

In [22]:
#check shape
renew.shape

(6916, 4)

In [23]:
renew.describe()

year  %_energy_renewables
count  6916.000000          6916.000000
mean   2002.500000             1.710929
std       7.500542             4.631711
min    1990.000000             0.000000
25%    1996.000000             0.000000
50%    2002.500000             0.000000
75%    2009.000000             1.053341
max    2015.000000            65.443742

In [26]:
#export wrangled renewables data
renew.to_csv(os.path.join(path,'02 data', 'prepared data', 'energy_from_renewables_wrangled.csv'))